### PR TITLE
[INFINITY-1988] Fix locking around suppress/revive

### PR DESCRIPTION
**Pull request by gabrielhartmann**

### CODE REVIEW

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/AbstractScheduler.java [hunk 1]
```diff
@@ -213,16 +216,46 @@ public void error(SchedulerDriver driver, String message) {
         SchedulerUtils.hardExit(SchedulerErrorCode.ERROR);
     }
 
+    protected void suppressOrRevive(PlanCoordinator planCoordinator) {
+        synchronized (suppressReviveLock) {
```
Comment (elezar) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/AbstractScheduler.java]: Question: I've seen `synchronized (this)` this used in some places online (e.g. http://winterbe.com/posts/2015/04/30/java8-concurrency-tutorial-synchronized-locks-examples/). Is there an advantage over using an explicit object other than readability (and the fact that one can only handle one lock with `synchronized (this)`?

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/AbstractScheduler.java [hunk 2]
```diff
@@ -213,16 +216,46 @@ public void error(SchedulerDriver driver, String message) {
...
+                } else {
+                    suppress();
+                }
+            }
+        }
+    }
+
     void suppress() {
         LOGGER.info("Suppressing offers.");
+        setOfferMode(true);
```
Comment (elezar) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/AbstractScheduler.java]: Should this not replace the following two lines?

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/AbstractScheduler.java (hunk 1 above)
Comment (gabrielhartmann) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/AbstractScheduler.java]: Both of the reasons you stated are the reasons.
1. Readability
1. Avoids accidentally having methods with different purposes using `this` to lock and unintentionally conflicting.

**Comment by gabrielhartmann** [comment]
@elezar:  Thanks for catching the failure to delete those lines.  Fixed.

**Comment by gabrielhartmann** [comment]
@elezar:  I doubt it, but it's possible.  Will have to investigate.

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/AbstractScheduler.java [hunk 3]
```diff
@@ -213,16 +216,44 @@ public void error(SchedulerDriver driver, String message) {
...
-        driver.reviveOffers();
-        StateStoreUtils.setSuppressed(stateStore, false);
+        setOfferMode(false);
+    }
+
+    private void setOfferMode(boolean suppressed) {
+        synchronized (suppressReviveLock) {
+            if (suppressed) {
+                LOGGER.info("Suppressing offers.");
+                driver.suppressOffers();
```
Comment (ArtRand) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/AbstractScheduler.java]: unhandled `StateStoreException` here yes?

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/AbstractScheduler.java [hunk 4]
```diff
@@ -213,16 +216,44 @@ public void error(SchedulerDriver driver, String message) {
...
+                StateStoreUtils.setSuppressed(stateStore, true);
+            } else {
+                LOGGER.info("Reviving offers.");
+                driver.reviveOffers();
```
Comment (ArtRand) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/AbstractScheduler.java]: and here

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/AbstractScheduler.java (hunk 3 above)
Comment (gabrielhartmann) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/AbstractScheduler.java]: Yes, by design.  If you can't write down your suppress/revive state, the Scheduler has no business staying up.

Comment (ArtRand) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/AbstractScheduler.java]: Yeah, I agree, maybe have a catch here to make it easier to debug?

Comment (gabrielhartmann) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/AbstractScheduler.java]: Would be inconsistent.  We could do that for every StateStore call, but we don't.  I don't know why I'd do it here and not everywhere else.

**Pull request merged**